source/framework/vx_array.c

```diff
@@ -22,6 +22,182 @@ static void ownInitArrayObject(
+static vx_status VX_CALLBACK arrayKernelCallback(vx_enum kernel_enum, vx_bool validate_only, vx_enum optimization, const vx_reference params[], vx_uint32 num_params);
+
+/*! \brief checks that the input and output references are not the same, and that both are valid arrays
+*/
+static vx_bool ownIsValidInputAndOutputArrays(vx_array input, vx_array output)
+{
+    if ((input != output) &&
+        (ownIsValidSpecificReference(&input->base, (vx_enum)VX_TYPE_ARRAY) == (vx_bool)vx_true_e) &&
+        (input->base.obj_desc != NULL) &&
+        (ownIsValidSpecificReference(&output->base, (vx_enum)VX_TYPE_ARRAY) == (vx_bool)vx_true_e) &&
+        (output->base.obj_desc != NULL)
+        )
+    {
+        return vx_true_e;
+    }
+    else
+    {
+        return vx_false_e;
+    }
+}
+
+/*! \brief This function is called to find out if it is OK to copy the input to the output.
+ * If the output is virtual, a zero capacity or invalid item type is OK
+ * Will also copy metadata from input to output.
+ * \returns VX_SUCCESS if it is, otherwise another error code.
+ *
+ */
+static vx_status isArrayCopyable(vx_array input, vx_array output)
+{
+    vx_status status = (vx_status)VX_SUCCESS;
+    tivx_obj_desc_array_t *ip_obj_desc = (tivx_obj_desc_array_t *)input->base.obj_desc;
+    tivx_obj_desc_array_t *op_obj_desc = (tivx_obj_desc_array_t *)output->base.obj_desc;
+    if (vx_false_e == ownIsValidInputAndOutputArrays(input, output))
+    {
+        status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+    }
+    else if (vx_true_e == output->base.is_virtual)
+    {
+        /* Either output type must be invalid or types must match*/
+        if ((VX_TYPE_INVALID != op_obj_desc->item_type) &&
+            (ip_obj_desc->item_type != op_obj_desc->item_type))
+        {
+            status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+        }
+        /* Either output capacity must be zero or at least as large as input capacity */
+        if ((0 != op_obj_desc->capacity) &&
+             (ip_obj_desc->capacity != op_obj_desc->capacity))
+        {
+            status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+        }
+    }
+    else if (op_obj_desc->item_type != ip_obj_desc->item_type)
+    {
+        /* Types must match */
+        status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+    }
+    else if (op_obj_desc->capacity < ip_obj_desc->capacity)
+    {
+        /* Output must have sufficient capacity to hold input */
+        status = (vx_status)VX_ERROR_NOT_COMPATIBLE;
+    }
+    else
+    {
+        op_obj_desc->item_type = ip_obj_desc->item_type;
+        op_obj_desc->item_size = ip_obj_desc->item_size;
+        op_obj_desc->capacity = ip_obj_desc->capacity;
+    }
+    return status;
+}
+
+/*! \brief This function is called to find out if it is OK to swap the input with the output.
+ * Both item type and capacity must be the same.
+ * \returns VX_SUCCESS if it is, otherwise another error code.
+ *
+ */
+static vx_status isArraySwapable(vx_array input, vx_array output)
```

Comment:
GENERIC: If you want to compare meta format between two references of the same type, we already have a generic function for this: tivxIsReferenceMetaFormatEqual. You can use this to simplify and generalize for all/most data types.

Reply:
Yes why not, but there are always some specific things to check for a specific object. I give you an example:
- vx image: you have to check if the output image is not an uniform  
- vx array: you have check if this is virtual
these checks are not in the function tivxIsReferenceMetaFormatEqual.
you want me to first check the meta with the tivxIsReferenceMetaFormatEqual and then check the specific things?

Reply:
tivxIsReferenceMetaFormatEqual
vx_lut copyable --> ok
vx_matrix --> origin and pattern is missing
vx_convolution --> mo meta comparision for convolution
vx_distrbution --> ok
vx_scalar --> ok
vx_tensor --> not all data checked like scaling_divisor_fixed_point_position
vx_pyramid --> specific checks if it is virtual
vx_image --> also too specific
vx_objarray --> ok, but questions in the code

swapable only two objects have a specific funtion
- vx_image --> cannot be use because of the subimage checked
- vx_array --> ok

copy function: (ownCopyReferenceGeneric)
vx_distrib: done
vx_array : done
vx_matrix: done
vx_lut: done
scalar: not needed
vx_tensor: done
vx_convolution: done
vx_distrbution: done
vx_pyramid: not applicable
vx_image / raw_image : not applicable
vx_objarray: not applicable

swap function:
vx_lut: done
vx_distrbution: done
vx_matrix: done
vx_array: done
vx:tensor: mixed with move, not possible
vx_convolution: done
vx_pyramid: not applicable
vx_image / raw_image : not applicable
vx_objarray: not applicable